### DISCUSSION TITLE
UI Idea: Combine Live and History view and add "zoom" and direct scrubbing for the timeline

### DISCUSSION THREAD

**Discussion by DReffects** [blakeblackshear/frigate#12255] [Ideas]
Coming from Unifi Protect I very much like their "one for all" user interface. 
When launching an individual camera view you are being thrown in the live stream but can instantly scrub down to previous recordings:
![image](https://github.com/blakeblackshear/frigate/assets/48361672/7a463af1-88c7-4044-8d26-961cc781a629)

**Why is this a usability plus?**
I often find myself hearing something rather loud or unexpected on the outside or see something vague through the blinds.
I open up the NVR but of course the event already has passed. 

Since by far not every audible event triggers motion in specifed areas (public areas for example) or creates detections it's rather complicated to at first get a live overview but then scrub back i time since you have to switch views back and forth. 
This would especially be handy on mobile.

**"direct scrubbing" for the timeline:**
It's kind of hard to explain but instead of first scrolling the timeline and then touching the actual time itself I would prefere if the scrolling itself would advance/reverse the shown video.

Also a pinch-zoom feature would be great for the timeline to focus on specific areas or "zoom out" to quickly scroll through a whole day to find something specific. 

(Very much a clone of the Unifi Protect UI ;-) )

I don't know what you guys think about the unifi protect UI - let me know if.

**Comment by NickM-27** [comment]
I would suggest reading [the blog post about the design of the UI](https://github.com/blakeblackshear/frigate/discussions/11136). Combining the screens was considered but ultimately decided against. 

> It's kind of hard to explain but instead of first scrolling the timeline and then touching the actual time itself I would prefere if the scrolling itself would advance/reverse the shown video.

you can already do this by just clicking (or tapping) the handlebar and dragging to scrub. When you get to the end of the timeline it will scroll for you

**Reply by NickM-27** [reply]
and zooming is pinned as a feature request https://github.com/blakeblackshear/frigate/issues/11398, feel free to add a 👍

**Reply by DReffects** [reply]
Thanks. Good to hear it was considered!

About the scrubbing: I've tried that with mixed results. The primary issue for me is the fixed speed at which the scrubbing takes place.
Having the red marker "fixed" and just scrolling the timeline would result in way less steps to reach a certain point in time, plus it eliminates the need to tap again into the timeline after scrolling. 

I am quite new to the whole github experience - what would be the best way to submit feature suggestions / UI ideas? Like this with a new topic with combined ideas and thoughts or via a dedicated topci for each suggestion/question?

Frigate is quite awesome and I'd like to give more input if you don't mind :-)

**Reply by NickM-27** [reply]
Scrolling is problematic for multiple reasons. It is technically more difficult, it makes it so you have to change the current time (where as right now you can still scroll the timeline to view the activity at a different point in time without selecting that new time and losing your current spot), and scrolling could result in other issues as well. 

You can always create a feature request as an issue but, for example this feature request would be closed because the UI is working the way it was designed and the way we intend for it to work.

**Reply by DReffects** [reply]
Thanks for the engaging conversation!

The request comes from my daily routine of fast-scrubbing through the nightly recordings (i record 24/7 for all cams) while being in bed in the morning. 
This works at the exact speed of my thumb with the unifi app and lets me granually scrub through a long period of recording with instant rewinds (thumb goes downwards), replays (thub goes upwards) and realtime playback (thumb removed from the screen). At least for me this is vastly superiour compared to the frigate method, that requires me to scroll at a fixed speed while dragging the handlebar. 

I absoluetely recognize that it works as intended and has a very strong use case for the situations you described. I just want to shed light on another way of using or "experiencing" recordings I find quite useful. In fact I would fancy both options, since the current way of scrubbing does not work well for my morning routine but has exellent applications for other scenarios when reviewing footage. 

I love the idea of the UI - I'd like to see the following optimizations:
- Way larger Date display (currently at 8px) in the timeline. It's simply too small for me to read (I am getting old^^)
- A prominent date seperator in the timeline would be super helpful. just a marker with the date at midnight
- I find the contrast in dark mode to be unreadable in most cases, especially in the timeline and stats pages (yeah, i am old)
- is there an option to enable 24h time format instead of am/pm ?
- And since frigate is so tightly integrated in Home Assistant: it would be great if the general font styling would be more like the home assistant default theme to make it more seamless (unless you want to stand out of course)

Before:
![image](https://github.com/blakeblackshear/frigate/assets/48361672/51d30924-7b09-463e-9d9b-fa57f4a8cad4)
![image](https://github.com/blakeblackshear/frigate/assets/48361672/ad943258-c225-4c9e-87ef-6be17c1dfb10)

After - Suggestion:
![image](https://github.com/blakeblackshear/frigate/assets/48361672/c29837df-668d-41d1-a049-704c038d38af)

I'd be happy to build a custom color scheme/theme if you could point me into the right direction.

**Reply by NickM-27** [reply]
> At least for me this is vastly superiour compared to the frigate method, that requires me to scroll at a fixed speed while dragging the handlebar.

the way you are describing it still makes it seem like you are not using it correctly. You can move the handlebar up and the further up it is moved the faster the timeline will scroll automatically.

> I absoluetely recognize that it works as intended and has a very strong use case for the situations you described. I just want to shed light on another way of using or "experiencing" recordings I find quite useful.

Feedback is appreciated, to be clear a lot of thought has been put in to how it works. Also in general, we have only received this feedback from a few users that are used to another software, and some agreed after getting used to the way frigate does it that they found it to be superior. 

> is there an option to enable 24h time format instead of am/pm ?

yes, this is inside of the config under `ui`

> After - Suggestion:

In your example the times cover important data on the timeline. It is possible some scaling or other visual changes mode for increased readability can be considered in a future version but I wouldn't expect it for 0.14.0

**Reply by NickM-27** [reply]
Themes are put inside of https://github.com/blakeblackshear/frigate/tree/dev/web/themes

**Reply by DReffects** [reply]
> the way you are describing it still makes it seem like you are not using it correctly. You can move the handlebar up and the further up it is moved the faster the timeline will scroll automatically.

Hmm maybe its a usability issue in relation to the screen size of my phone. I have a rather large screen (Xiaomi 14 Ultra)
I've just noticed, thanks to your explanation, that there is a very small area on the top and bottom of the timeline where i have some sort of gradual control over the scroll speed. 

But that area is like 2-3 milimeters on my phone screen and the vertical area of the timeline is too tall to even reach the top with my thumb when holding the phone. That's why I have not noticed before.
This is where my hassle comes from, since I cannot quickly move back and forth. 

I sort of have to hit the gradual scroll area pixel perfect to not reach terminal scroll velocity and reposition the phone in my hand to reach the top of the timeline with my thumb. This is probably a non-issue with smaller phone screens. 

If i enable "one hand mode" which shrinks the vertical space of the screen, the timeline does not work properly anymore because the viewport does not "refresh" to reflect the new screen height. The bottom navigation in the history view is missing, therfor I cannot use scroll at all. 

Here's a screenshot with reduced vertical screen height:
![Screenshot_2024-07-02-22-25-05-576_io homeassistant companion android](https://github.com/blakeblackshear/frigate/assets/48361672/c6c6363c-79b5-42a6-90ea-a01db135566a)

If I am on the other hand just scroll the timeline and want scroll to stop i have to tap the screen unless i want to wait since it slowly stops itself (at the wrong position^^). then the handlebar jumps to the point of the tap that was used to stop the scrolling. 

Here's a crude recording:

https://github.com/blakeblackshear/frigate/assets/48361672/5a5728f4-edcc-4e6c-b701-0bc43251150d

https://github.com/blakeblackshear/frigate/assets/48361672/d3882b80-7e17-41d0-889d-ba4a294edc1a

> Feedback is appreciated, to be clear a lot of thought has been put in to how it works. 
Awesome!

> Also in general, we have only received this feedback from a few users that are used to another software, and some agreed after getting used to the way frigate does it that they found it to be superior.

As you might have noticed, english is not my primary language. All feedback is being made in good faith and should please not be recieved as "ungrateful" or with an unwillingness to change systems. 
Perhaps my personal use case is not that common, too. I find myself scrolling through the nightly recordings of my toddlers' cam in the morning without much granual control, just swiping the thumb upwards on the unifi protect timeline, enjoying the movements. 

I very much recognize that the frigate handlebar/timeline combination offers more control in certain situation. I find myself quite struggeling with the unifi UI in that regard, if i want to locate an exact point in time up to the frame. Your handlebar allows that :)

> yes, this is inside of the config under ui
Thanks!!

> In your example the times cover important data on the timeline.

Yes :-( A suggestion would be to make the timeline in desktop view double in width with the actual timecodes on the one side and the motion information on the other. there's a lot of empty space there anyway:

![image](https://github.com/blakeblackshear/frigate/assets/48361672/946eb5a2-1b14-469c-b6a0-7a6145044b12)

> Themes are put inside of https://github.com/blakeblackshear/frigate/tree/dev/web/themes

Cool! I will take a look. How can I edit this on the Home assistant file system?
I was not able to find the css files when accessing HA via SSH.

**Reply by NickM-27** [reply]
> As you might have noticed, english is not my primary language.

actually no, I could not really tell 👍 

> If i enable "one hand mode" which shrinks the vertical space of the screen, the timeline does not work properly anymore because the viewport does not "refresh" to reflect the new screen height. The bottom navigation in the history view is missing, therfor I cannot use scroll at all.

unfortunately the "one hand mode" just shifts the window downwards, it does not tell the browser to resize the window so we can't really adapt to this.
